[Temporal] Bounds-check year in balanceISODate()

catamorphism opened the pull request:
#### 9b21b88466cfab82d46e76e545109ed0e1bf0e78
<pre>
[Temporal] Bounds-check year in balanceISODate()
<a href="https://bugs.webkit.org/show_bug.cgi?id=301659">https://bugs.webkit.org/show_bug.cgi?id=301659</a>

Reviewed by Yusuke Suzuki and Sosuke Suzuki.

balanceISODate() wasn&apos;t checking for the invalid-year case, which
caused it in some cases to turn an invalid date into a valid one
(e.g. by adding months).

* JSTests/stress/temporal-plaindate.js:
(shouldBe):
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::balanceISODate):

Canonical link: <a href="https://commits.webkit.org/302413@main">https://commits.webkit.org/302413@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ecfac17be6f5e7c776e99945194050b4fe5d8251

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1127 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39702 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136253 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80236 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1078 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1005 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98107 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66020 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131817 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/820 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115433 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78719 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/750 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33544 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79532 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/120880 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109188 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34038 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138714 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127329 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/943 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/913 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106644 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/997 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111767 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106457 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27137 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/767 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30299 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/53400 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1012 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64321 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160341 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/855 "Built successfully") | | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40017 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; jscore-tests (failure); Compiled JSC; jscore-tests; Found unexpected failure with change (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/911 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/947 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -251,6 +251,9 @@ ISO8601::PlainDate TemporalCalendar::isoDateFromFields(JSGlobalObject* globalObj
 ISO8601::PlainDate TemporalCalendar::balanceISODate(JSGlobalObject* globalObject, double year, double month, double day)
 {
+    // Avoid turning an out-of-range date into an in-range date
+    if (static_cast<int32_t>(year) == ISO8601::outOfRangeYear) [[unlikely]]
```
`static_cast` for `Infinity` or `NaN` to `int32_t` is undefined behavior. If `year` is guaranteed not to be these values, I think an `ASSERT` to verify that should be added. If `year` can be these values, it needs to be handled appropriately.

catamorphism replied:
Temporal properties are always finite, so I added an assertion to that effect.

sosukesuzuki replied:
Also should add `std::isnan` assertion?

catamorphism replied:
I thought `std::isfinite` returns false for `NaN`? https://en.cppreference.com/w/cpp/numeric/math/isfinite

sosukesuzuki replied:
Ooops! I didn't know that, thanks!

sosukesuzuki reviewed (approved):
r=me

catamorphism committed:
[Temporal] Bounds-check year in balanceISODate()
https://bugs.webkit.org/show_bug.cgi?id=301659

Reviewed by Yusuke Suzuki and Sosuke Suzuki.

balanceISODate() wasn't checking for the invalid-year case, which
caused it in some cases to turn an invalid date into a valid one
(e.g. by adding months).

* JSTests/stress/temporal-plaindate.js:
(shouldBe):
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::balanceISODate):

Canonical link: https://commits.webkit.org/302413@main
